### PR TITLE
Use std::exchange in move operator for Tensors

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Use the std::exchange in move operators,
to have single statement:
```x_ = std::exchange(other.x_, 0);```
instead of multiple separate lines:
```
x_ = other.x_;
other.x_ = 0;
``` 

This adjusts Tensor and SampleView types,
so they both have the empty {0} shape after
move.
Previously it left the Tensor with a scalar shape
that could indicate existing allocation after
move.

As the use after move shouldn't be practised,
this change should not impact anything.

TensorViews and TensorShapes are left mostly
unaffected as the static_dim doesn't compose
nicely with forcing {0} shape or clearing
the dimensionality. For a TensorView, it is
valid to have a shape and a nullptr,
for the shapes they can have dimensionality
and no samples.

### Affected modules and functionalities:
Tensor, SampleView.

### Tests:

- [x] Existing tests apply
Sample View already has std::move tests.
TensorListShape tests invoke the move constructor.
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2966

### CODE REVIEW

**Comment by klecki** [comment]
!build

**Review by prak-nv — APPROVED** [review]

File: include/dali/core/tensor_shape.h [hunk 1]
```diff
@@ -795,9 +795,8 @@ struct TensorListShape<DynamicDimensions>
   TensorListShape &operator=(const TensorListShape &) = default;
   TensorListShape &operator=(TensorListShape &&other) {
     shapes = std::move(other.shapes);
-    nsamples = other.size();
-    ndim = other.sample_dim();
-    other.nsamples = 0;
+    nsamples = std::exchange(other.nsamples, 0);
+    ndim = other.sample_dim();  // Sample dim could be static, and it is valid to not zero it.
```
Comment (szalpal) [inline include/dali/core/tensor_shape.h:799]: IIUC, previously the `nsamples` wasn't moved, but passed by value, so this changes the behaviour. Is that intentional?

File: include/dali/core/tensor_shape.h [hunk 2]
```diff
@@ -812,9 +811,8 @@ struct TensorListShape<DynamicDimensions>
   template <int other_sample_ndim>
   TensorListShape &operator=(TensorListShape<other_sample_ndim> &&other) {
     shapes = std::move(other.shapes);
-    nsamples = other.size();
-    ndim = other.sample_dim();
-    other.nsamples = 0;
+    nsamples = std::exchange(other.nsamples, 0);
+    ndim = other.sample_dim();  // Sample dim could be static, and it is valid to not zero it.
```
Comment (szalpal) [inline include/dali/core/tensor_shape.h:815]: IIUC, previously the nsamples wasn't moved, but passed by value, so this changes the behaviour. Is that intentional?

File: include/dali/core/tensor_shape.h [hunk 3]
```diff
@@ -894,8 +892,7 @@ struct TensorListShape : TensorListShapeBase<TensorListShape<sample_ndim>, sampl
   TensorListShape &operator=(const TensorListShape &) = default;
   TensorListShape &operator=(TensorListShape &&other) {
     shapes = std::move(other.shapes);
-    nsamples = other.size();
-    other.nsamples = 0;
+    nsamples = std::exchange(other.nsamples, 0);
```
Comment (szalpal) [inline include/dali/core/tensor_shape.h:895]: IIUC, previously the nsamples wasn't moved, but passed by value, so this changes the behaviour. Is that intentional?

**Review by szalpal — APPROVED** [review]

File: include/dali/core/tensor_shape.h (hunk 1 above)
Comment (klecki) [reply include/dali/core/tensor_shape.h:799]: It doesn't matter if it's moved, it's an int.

File: include/dali/core/tensor_shape.h (hunk 2 above)
Comment (klecki) [reply include/dali/core/tensor_shape.h:815]: As above.

File: include/dali/core/tensor_shape.h (hunk 3 above)
Comment (klecki) [reply include/dali/core/tensor_shape.h:895]: As above

File: include/dali/core/tensor_shape.h (hunk 1 above)
Comment (prak-nv) [reply include/dali/core/tensor_shape.h:799]: it is basically `int nsamples; nsamples = int::int(static_cast<int&&>(other.nsamples)); other.nsamples = 0; ` where both int copy and move constructors are trivial (ie. `memcpy`).

**Pull request merged**